c/ls-clone/babyls.c

```diff
@@ -0,0 +1,258 @@
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <dirent.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <pwd.h>
+#include <grp.h>
+#include <time.h>
+
+#define MAX_DIRECTORY_SIZE 100
+
+// next implement long format
+// stretch implement colors
+
+// permissions mask values for st_mode organized
+// in order of long format display
+int permission_lookup[] = {
+	S_IRUSR, S_IWUSR, S_IXUSR,
+	S_IRGRP, S_IWGRP, S_IXGRP,
+	S_IROTH, S_IWOTH, S_IXOTH
+};
+
+// character lookup table for permission_lookup masks
+char permission_char[] = {
+	'r', 'w', 'x',
+	'r', 'w', 'x',
+	'r', 'w', 'x'
+};
+
+// file type masks
+// saw this after making this table
+// but man 7 inode shows you can do pre-made macros like S_ISDIR(m)
+int file_type_lookup[] = {
+	S_IFMT, S_IFSOCK, S_IFLNK,
+	S_IFREG, S_IFBLK, S_IFDIR,
+	S_IFCHR, S_IFIFO
+};
+
+// file type characters
+// should add support beyond directory
+char file_type_char[] = {
+	'-', '-', '-',
+	'-', '-', 'd',
+	'-', '-'
+};
+
+int alpha_dirbuf_comp(const void *d1, const void *d2) {
+	struct dirent *dirbuf1 = (struct dirent *)d1;
+	struct dirent *dirbuf2 = (struct dirent *)d2;
+	return strcmp(dirbuf1->d_name, dirbuf2->d_name);
+}
+
+// could be generalized to all comp funcs if new sorting is added
+int r_alpha_dirbuf_comp(const void *d1, const void *d2) {
+	int res = alpha_dirbuf_comp(d1, d2);
+	return res * -1;
+}
+
+void print_directories_names(struct dirent dirarr[], int len) {
+    	int i;
+    	for (i = 0; i < len; i++) {
+    		printf("%s  ", dirarr[i].d_name); 
+    	}
+    	printf("\n");
+}
+
+char* permissions_str(unsigned long st_mode) {
+	// interestingly if we try to use the stack
```

Comment:
I tried to reproduce the "working" string on the stack but couldn't. Didn't break up commits enough to see what I did originally but this doesn't accidentally work in this example:
```
john@tank:~/csi/src/johndavidmartinez/c/ls-clone$ cat stackin.c
#include <stdio.h>

char* first() {
    int i;
    char first[10];
    for (i = 0; i < 10; i++) {
        first[i] = 'o';
    }
    first[9] = '\0';
    return first;
}

int main() {
    char* ok = first();
    printf("%s\n", ok);
}
john@tank:~/csi/src/johndavidmartinez/c/ls-clone$ cc stackin.c
stackin.c:10:12: warning: address of stack memory associated with local variable 'first' returned [-Wreturn-stack-address]
    return first;
           ^~~~~
1 warning generated.
john@tank:~/csi/src/johndavidmartinez/c/ls-clone$ ./a.out

john@tank:~/csi/src/johndavidmartinez/c/ls-clone$
```

Now with the example you suggested
```
john@tank:~/csi/src/johndavidmartinez/c/ls-clone$ cat stackin_e.c
#include <stdio.h>

void first(char* buf) {
    int i;
    for(i = 0; i < 4; i++) {
        buf[i] = 'f';
    }
    buf[3] = '\0';
}

void second(char* buf) {
    int i;
    for(i = 0; i < 4; i++) {
        buf[i] = 's';
    }
    buf[3] = '\0';
}

int main() {
    char f[4], s[4];
    first(f);
    second(s);

    printf("%s\n", f);
    printf("%s\n", s);
}
john@tank:~/csi/src/johndavidmartinez/c/ls-clone$ ./a.out
fff
sss
john@tank:~/csi/src/johndavidmartinez/c/ls-clone$
```
Two character arrays are initialized on the stack in main. Their pointers are passed to `first` and `second` and we don't need to call & because arrays decay into pointers so it is implicit that `f` and `s` in this case point to the memory address of the first element. We assign character terminated by a null byte and so printf works as expected printing the string.

Makes sense but not sure how to reproduce the (undefined?) behavior I saw before.

As I was writing this it occured to me to try this:
```
john@tank:~/csi/src/johndavidmartinez/c/ls-clone$ cat stackin_nested.c
#include <stdio.h>

char* first() {
    int i;
    char ok[4];
    for (i = 0; i < 4; i++) {
        ok[i] = 'z';
    }
    ok[3] = '\0';
    return ok;
}

void nest() {
    printf("%s\n", first());
}

int main() {
    nest();
}
john@tank:~/csi/src/johndavidmartinez/c/ls-clone$ cc stackin_nested.c
stackin_nested.c:10:12: warning: address of stack memory associated with local variable 'ok' returned [-Wreturn-stack-address]
    return ok;
           ^~
1 warning generated.
john@tank:~/csi/src/johndavidmartinez/c/ls-clone$ ./a.out
N&
john@tank:~/csi/src/johndavidmartinez/c/ls-clone$```

Here we're retrying to reproduce the behavior but now we're another later deep in the call stack and we do get something, random bytes. So it could be in the original scenario I got lucky and the undefined behavior just happened to work or I got confused I _thought_ it was working. 😅 

To bring it home. The stack gets "freed" once we return from the function. In your example we're still in the same point in the stack so our allocated buffer `buf` is still there.

Its not quite right but my intuition is to think of it as "scope" we're out of the "scope" in my two bad examples. My guess is I can fix this intuition by reviewing stack memory.

Reply:
Right, once you return from a function, any stack space can be used (clobbered) by subsequent function calls. You might get lucky and your data is still there (unclobbered), but you can't count on that.